Panic on verifying platform advanceTimeTx

**Describe the bug**
Node crashes with:

```

panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x962d36]

goroutine 134 [running]:
github.com/ava-labs/gecko/utils/logging.(*Log).StopOnPanic(0xc0007f1440)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/utils/logging/log.go:249 +0x11d
panic(0xd3a7c0, 0x1914b50)
        /usr/local/go/src/runtime/panic.go:967 +0x166
github.com/ava-labs/gecko/database/versiondb.(*Database).Has(0x0, 0xc000370f20, 0x20, 0x20, 0xc0000ffe00, 0x0, 0x0)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/database/versiondb/db.go:40 +0x46
github.com/ava-labs/gecko/vms/components/state.(*state).Get(0xc000033c30, 0x10f71e0, 0x0, 0xfffffffffffffffd, 0xc0000386a0, 0x10000000106e12c, 0xc0000f5418, 0x504612, 0xc0004f83b8)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/vms/components/state/state.go:132 +0xc8
github.com/ava-labs/gecko/vms/components/state.(*state).GetTime(0xc000033c30, 0x10f71e0, 0x0, 0xc0000386a0, 0x14c5b85400e23200, 0x5ea90daa, 0xc0000f53b0, 0x4e3436, 0x4e2b60)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/vms/components/state/state.go:197 +0x56
github.com/ava-labs/gecko/vms/platformvm.(*VM).getTimestamp(0xc000135200, 0x10f71e0, 0x0, 0xc0000f54a0, 0x4ed5a5, 0xc000140060, 0xc0004f8300, 0xc6)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/vms/platformvm/state.go:188 +0x5c
github.com/ava-labs/gecko/vms/platformvm.(*advanceTimeTx).SemanticVerify(0xc001b82190, 0x10f71e0, 0x0, 0x7f7c600d27d0, 0xc000e1e690, 0x180000000097dd01, 0x4, 0xc0008ed3b0, 0xc0000f55a0)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/vms/platformvm/advance_time_tx.go:55 +0x9c
github.com/ava-labs/gecko/vms/platformvm.(*ProposalBlock).Verify(0xc000f1c230, 0xc0000f5690, 0x1)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/vms/platformvm/proposal_block.go:106 +0xa2
github.com/ava-labs/gecko/snow/engine/snowman.(*Transitive).deliver(0xc000270000, 0x10f1ec0, 0xc000f1c230)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/snow/engine/snowman/transitive.go:350 +0x10b
github.com/ava-labs/gecko/snow/engine/snowman.(*issuer).Update(...)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/snow/engine/snowman/issuer.go:43
github.com/ava-labs/gecko/snow/engine/snowman.(*issuer).Fulfill(0xc0004e8630, 0xc00051cc80)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/snow/engine/snowman/issuer.go:22 +0xa8
github.com/ava-labs/gecko/snow/events.(*Blocker).Fulfill(0xc000270278, 0xc00051cc80)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/snow/events/blocker.go:32 +0x161
github.com/ava-labs/gecko/snow/engine/snowman.(*Transitive).deliver(0xc000270000, 0x10f1ec0, 0xc000f1c310)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/snow/engine/snowman/transitive.go:384 +0x36b
github.com/ava-labs/gecko/snow/engine/snowman.(*issuer).Update(...)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/snow/engine/snowman/issuer.go:43
github.com/ava-labs/gecko/snow/engine/snowman.(*issuer).Fulfill(0xc0004e8810, 0xc000932780)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/snow/engine/snowman/issuer.go:22 +0xa8
github.com/ava-labs/gecko/snow/events.(*Blocker).Fulfill(0xc000270278, 0xc000932780)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/snow/events/blocker.go:32 +0x161
github.com/ava-labs/gecko/snow/engine/snowman.(*Transitive).deliver(0xc000270000, 0x10f1e00, 0xc000c00900)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/snow/engine/snowman/transitive.go:379 +0x2b2
github.com/ava-labs/gecko/snow/engine/snowman.(*issuer).Update(0xc0004e8990)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/snow/engine/snowman/issuer.go:43 +0x61
github.com/ava-labs/gecko/snow/events.(*Blocker).Register(0xc000270278, 0x10e8720, 0xc0004e8990)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/snow/events/blocker.go:59 +0x286
github.com/ava-labs/gecko/snow/engine/snowman.(*Transitive).insert(0xc000270000, 0x10f1e00, 0xc000c00900)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/snow/engine/snowman/transitive.go:286 +0x183
github.com/ava-labs/gecko/snow/engine/snowman.(*Transitive).insertFrom(0xc000270000, 0xc000498600, 0x10f1e00, 0xc000c00900, 0x10f1e00)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/snow/engine/snowman/transitive.go:245 +0xc2
github.com/ava-labs/gecko/snow/engine/snowman.(*Transitive).Put(0xc000270000, 0xc000498600, 0x0, 0xc00062a180, 0xc0000301c8, 0x132, 0x132)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/snow/engine/snowman/transitive.go:102 +0x277
github.com/ava-labs/gecko/snow/networking/handler.(*Handler).dispatchMsg(0xc000b897a0, 0x8, 0xc000498600, 0x0, 0xc00062a180, 0xc0000301c8, 0x132, 0x132, 0x0, 0x0, ...)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/snow/networking/handler/handler.go:83 +0x42f
github.com/ava-labs/gecko/snow/networking/handler.(*Handler).Dispatch(0xc000b897a0)
        /home/adasri/go-workspace/src/github.com/ava-labs/gecko/snow/networking/handler/handler.go:43 +0x2b0
```

**To Reproduce**
Unknown. User says ` i left the node for some time (for tx to complete) and hit the api. then panic'ed.`

**Expected behavior**
Node doesn't crash ;)

**Screenshots**
See log above

**Operating System**
Ubuntu 18.04

**Additional context**
Thanks @adasari for reporting this bug

By submitting this issue I agree to the [Terms and Conditions](https://files.avalabs.org/legal/bug_hunt_q2_2020_terms.pdf) of the [Developer Accelerator Program](https://avalabs.org/dap).

--- a/vms/platformvm/abort_block.go
+++ b/vms/platformvm/abort_block.go
@@ -20,12 +20,13 @@ type Abort struct {
 //
 // This function also sets onAcceptDB database if the verification passes.
 func (a *Abort) Verify() error {
+	parent, ok := a.parentBlock().(*ProposalBlock)
 	// Abort is a decision, so its parent must be a proposal
-	if parent, ok := a.parentBlock().(*ProposalBlock); ok {
-		a.onAcceptDB, a.onAcceptFunc = parent.onAbort()
-	} else {
+	if !ok {
 		return errInvalidBlockType
 	}
+
+	a.onAcceptDB, a.onAcceptFunc = parent.onAbort()
 
 	a.vm.currentBlocks[a.ID().Key()] = a
 	a.parentBlock().addChild(a)

--- a/vms/platformvm/commit_block.go
+++ b/vms/platformvm/commit_block.go
@@ -21,11 +21,12 @@ type Commit struct {
 // This function also sets the onCommit databases if the verification passes.
 func (c *Commit) Verify() error {
 	// the parent of an Commit block should always be a proposal
-	if parent, ok := c.parentBlock().(*ProposalBlock); ok {
-		c.onAcceptDB, c.onAcceptFunc = parent.onCommit()
-	} else {
+	parent, ok := c.parentBlock().(*ProposalBlock)
+	if !ok {
 		return errInvalidBlockType
 	}
+
+	c.onAcceptDB, c.onAcceptFunc = parent.onCommit()
 
 	c.vm.currentBlocks[c.ID().Key()] = c
 	c.parentBlock().addChild(c)

--- a/vms/platformvm/common_blocks.go
+++ b/vms/platformvm/common_blocks.go
@@ -124,10 +124,6 @@ type CommonBlock struct {
 	*core.Block `serialize:"true"`
 	vm          *VM
 
-	// This block's parent.
-	// nil before parentBlock() is called on this block
-	parent Block
-
 	// This block's children
 	children []Block
 }
@@ -142,7 +138,6 @@ func (cb *CommonBlock) Reject() {
 // free removes this block from memory
 func (cb *CommonBlock) free() {
 	delete(cb.vm.currentBlocks, cb.ID().Key())
-	cb.parent = nil
 	cb.children = nil
 }
 
@@ -165,11 +160,6 @@ func (cb *CommonBlock) Parent() snowman.Block {
 
 // parentBlock returns this block's parent
 func (cb *CommonBlock) parentBlock() Block {
-	// Check if the block already has a reference to its parent
-	if cb.parent != nil {
-		return cb.parent
-	}
-
 	// Get the parent from database
 	parentID := cb.ParentID()
 	parent, err := cb.vm.getBlock(parentID)
@@ -177,7 +167,6 @@ func (cb *CommonBlock) parentBlock() Block {
 		cb.vm.Ctx.Log.Debug("could not get parent (ID %s) of block %s", parentID, cb.ID())
 		return nil
 	}
-	cb.parent = parent
 	return parent.(Block)
 }
 

--- a/vms/platformvm/proposal_block.go
+++ b/vms/platformvm/proposal_block.go
@@ -92,15 +92,18 @@ func (pb *ProposalBlock) onAbort() (*versiondb.Database, func()) {
 //
 // If this block is valid, this function also sets pas.onCommit and pas.onAbort.
 func (pb *ProposalBlock) Verify() error {
-	// pdb is the database if this block's parent is accepted
-	var pdb database.Database
-	parent := pb.parentBlock()
+	parentIntf := pb.parentBlock()
+
 	// The parent of a proposal block (ie this block) must be a decision block
-	if parent, ok := parent.(decision); ok {
-		pdb = parent.onAccept()
-	} else {
+	parent, ok := parentIntf.(decision)
+	if !ok {
 		return errInvalidBlockType
 	}
+
+	// pdb is the database if this block's parent is accepted
+	pdb := parent.onAccept()
+
+	pb.vm.Ctx.Log.Warn("Got %+v from %p", pdb, parentIntf)
 
 	var err error
 	pb.onCommitDB, pb.onAbortDB, pb.onCommitFunc, pb.onAbortFunc, err = pb.Tx.SemanticVerify(pdb)
@@ -109,7 +112,7 @@ func (pb *ProposalBlock) Verify() error {
 	}
 
 	pb.vm.currentBlocks[pb.ID().Key()] = pb
-	parent.addChild(pb)
+	parentIntf.addChild(pb)
 	return nil
 }
 

--- a/vms/platformvm/vm_test.go
+++ b/vms/platformvm/vm_test.go
@@ -1596,3 +1596,81 @@ func TestBootstrapPartiallyAccepted(t *testing.T) {
 
 	router.Shutdown()
 }
+
+func TestUnverifiedParent(t *testing.T) {
+	genesisAccounts := GenesisAccounts()
+	genesisValidators := GenesisCurrentValidators()
+	genesisChains := make([]*CreateChainTx, 0)
+
+	genesisState := Genesis{
+		Accounts:   genesisAccounts,
+		Validators: genesisValidators,
+		Chains:     genesisChains,
+		Timestamp:  uint64(defaultGenesisTime.Unix()),
+	}
+
+	genesisBytes, err := Codec.Marshal(genesisState)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	db := memdb.New()
+
+	vm := &VM{
+		SnowmanVM:    &core.SnowmanVM{},
+		chainManager: chains.MockManager{},
+	}
+
+	defaultSubnet := validators.NewSet()
+	vm.validators = validators.NewManager()
+	vm.validators.PutValidatorSet(DefaultSubnetID, defaultSubnet)
+
+	vm.clock.Set(defaultGenesisTime)
+	ctx := defaultContext()
+	msgChan := make(chan common.Message, 1)
+	if err := vm.Initialize(ctx, db, genesisBytes, msgChan, nil); err != nil {
+		t.Fatal(err)
+	}
+
+	firstAdvanceTimeTx, err := vm.newAdvanceTimeTx(defaultGenesisTime.Add(time.Second))
+	if err != nil {
+		t.Fatal(err)
+	}
+	firstAdvanceTimeBlk, err := vm.newProposalBlock(vm.Preferred(), firstAdvanceTimeTx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	vm.clock.Set(defaultGenesisTime.Add(2 * time.Second))
+	if err := firstAdvanceTimeBlk.Verify(); err != nil {
+		t.Fatal(err)
+	}
+
+	options := firstAdvanceTimeBlk.Options()
+	firstOption := options[0]
+	secondOption := options[1]
+
+	secondAdvanceTimeTx, err := vm.newAdvanceTimeTx(defaultGenesisTime.Add(2 * time.Second))
+	if err != nil {
+		t.Fatal(err)
+	}
+	secondAdvanceTimeBlk, err := vm.newProposalBlock(firstOption.ID(), secondAdvanceTimeTx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	parentBlk := secondAdvanceTimeBlk.Parent()
+	if parentBlkID := parentBlk.ID(); !parentBlkID.Equals(firstOption.ID()) {
+		t.Fatalf("Wrong parent block ID returned")
+	}
+
+	if err := firstOption.Verify(); err != nil {
+		t.Fatal(err)
+	}
+	if err := secondOption.Verify(); err != nil {
+		t.Fatal(err)
+	}
+	if err := secondAdvanceTimeBlk.Verify(); err != nil {
+		t.Fatal(err)
+	}
+}

